src/ucs/sys/module.c

```diff
@@ -115,6 +115,39 @@ static void ucs_module_loader_init_paths()
     }
 }
 
+/* Perform shallow search for a symbol */
+static void *ucs_module_dlsym_shallow(const char *module_path, void *dl, const char *symbol)
+{
+    Dl_info dl_info;
+    void *addr;
+    int ret;
+
+    addr = dlsym(dl, symbol);
+    if (addr == NULL) {
+        ucs_module_trace("could not find symbol '%s' in %s", symbol, module_path);
+        return NULL;
+    }
+
+    (void)dlerror();
+    ret = dladdr(addr, &dl_info);
+    if (ret == 0) {
+        ucs_module_debug("dladdr(%p) [%s] failed: %s", addr, symbol, dlerror());
+        return NULL;
+    }
+
+    /* return the symbol only if it was found in the requested library, and not,
+     * for example, in one of its dependencies.
+     */
+    if (strcmp(dl_info.dli_fname, module_path)) {
```

Comment:
minor: does it guarantee that dl_info.dli_fname has absolute path? even in case if it points to symbolic link?

Reply:
good point... i will check if we can compare something more strict than the path

Reply:
readlink + fstat -> compare inodes